src/query/interpret/eval.hpp

```diff
@@ -159,50 +224,54 @@ class ExpressionEvaluator : public ExpressionVisitor<TypedValue> {
   }
 
   TypedValue Visit(SubscriptOperator &list_indexing) override {
-    auto lhs = list_indexing.expression1_->Accept(*this);
+    ReferenceExpressionEvaluator referenceExpressionEvaluator(frame_, symbol_table_, ctx_);
+
+    TypedValue *lhs_p = nullptr;
+    lhs_p = list_indexing.expression1_->Accept(referenceExpressionEvaluator);
+    TypedValue lhs;
+    bool referenced = lhs_p != nullptr;
+    if (!lhs_p) {
+      lhs = list_indexing.expression1_->Accept(*this);
+      lhs_p = &lhs;
+    }
```

Comment:
This look a bit messy in my opinion, a few proposals:
```suggestion
    ReferenceExpressionEvaluator referenceExpressionEvaluator(frame_, symbol_table_, ctx_);

    TypedValue *lhs_ptr = list_indexing.expression1_->Accept(referenceExpressionEvaluator);
    TypedValue lhs;
    if (nullptr == lhs_ptr) {
      lhs = list_indexing.expression1_->Accept(*this);
      lhs_ptr = &lhs;
    }
```

And of course change `lhs_p` to `lhs_ptr` in the rest of the function.